File: PythonPlusR/data-cleaning.py
import pandas as pd
import numpy as np
import sys
import os
import enum
import matplotlib.pyplot as plt
import matplotlib
 
matplotlib.rcParams['font.size'] = 10
matplotlib.rcParams['figure.dpi'] = 100
 
from IPython.core.pylabtools import figsize

class EjectionType(enum.Enum):
    iqr = 0
    hampel = 1

def clean_table(df:pd.DataFrame, 
                path:str, 
                nan_crit_val:float = 0.25,
                iqr_coef:int = 1,
                ejection_type:EjectionType = EjectionType.iqr, #https://leftjoin.ru/all/outliers-detection-in-python/
                save_csv:bool = False, 
                do_output:bool = False):
    df = df.copy()
    log = []
    print('START data cleaning')

    for c in df.columns:
        if not pd.api.types.is_numeric_dtype(df[c]):
            df[c] = df[c].astype(str)

    # Удаление дубликатов
    df.drop_duplicates()

    # Обработка выбросов
    for c in df.columns:
        if pd.api.types.is_numeric_dtype(df[c]):
            median = df[c].median()

            if (ejection_type == EjectionType.iqr):
                Q1 = df[c].quantile(0.25)
                Q3 = df[c].quantile(0.75)
                max_deviation = (Q3 - Q1)*iqr_coef
            elif (ejection_type == EjectionType.hampel):
                difference = np.abs(median-df[c])
                max_deviation = 3 * difference.median()

            maxv = get_correct_max(df[c], max_deviation, median)
            minv = get_correct_min(df[c], max_deviation, median)

            for i in range(df[c].count()):
                v = df[c].loc[i]
                if v > median + max_deviation:
                    log.append(f'in col {c} in row {i} top ej-n = {v} repl-d by {maxv}; md={max_deviation}, median={median}')
                    df.at[i, c] = maxv
                elif v < median - max_deviation:
                    log.append(f'in col {c} in row {i} bottom ej-n = {v} repl-d by {minv}; md={max_deviation}, median={median}')
                    df.at[i, c] = minv

    # Обработка пропущенных значений
    for c in df.columns:
        if df[c].isna().sum() > df[c].count() * nan_crit_val:
            log.append(f'Drop column "{c}" beacuse too much nans')
            df = df.drop([c], axis=1)
        elif pd.api.types.is_numeric_dtype(df[c]):
            df[c].fillna(value=df[c].mean(), inplace=True)
        else:
            df[c].fillna(value=df[c].mode(), inplace=True)

    print('END data cleaning')
    if save_csv:
        df.to_csv(os.path.join(path, r'res.csv'))

    if do_output:
        with open(f"{path}\\temp\\log.txt", 'w') as fp:
            fp.write('\n'.join(log))
    return df
    
def get_correct_max(col, max_deviation, median):
    maxv = sys.float_info.min
    for v in col:
        if v>maxv:
            dif = abs(median - v)
            if dif <= max_deviation:
                maxv = v
    return maxv

def get_correct_min(col, max_deviation, median):
    minv = sys.float_info.max
    for v in col:
        if v<minv:
            dif = abs(median - v)
            if dif <= max_deviation:
                minv = v
    return minv

def make_graphs(orig:pd.DataFrame, clear:pd.DataFrame):
    cols = clear.columns

    figsize(6, 4)

    for c in cols:
        plt.hist([orig[c], clear[c]])

def main():
    dict = {'First Score':[100, 90, np.nan, 95, 1, 56],
        'Second Score': [30, np.nan, 45, 56, -11111, 56],
        'Third Score':[52, 1240, 80, 98, 1, 56],
        'Fourth Score':[np.nan, np.nan, np.nan, 65, 1, 56],
        'STRS':['52', '40', '80', None, 'aaa', 56],}
    df = pd.DataFrame(dict)
    directory = os.getcwd()
    df = pd.read_csv(directory + '\\datasets\\avocado.csv')
    print(df)

    df1 = clean_table(df, 
                      directory, 
                      do_output=True, 
                      nan_crit_val = 0.0000000001,
                      ejection_type=EjectionType.iqr,
                      iqr_coef = 1000)
    print(df1)

    make_graphs(df,df1)

if __name__ == '__main__':
    main()

File: PythonPlusR/data-cleaning.r
library(ggplot2)

remove_ejections <- function(df, iqr_coef, ejection_type) {
    iqr_upper <- function(x, coef) {
        median <- median(x)
        Q1 <- quantile(x, probs = .25)
        Q3 <- quantile(x, probs = .75)
        iqr <- Q3 - Q1

        upper_limit <- median + (iqr * coef)

        return(x > upper_limit)
    }

    iqr_lower <- function(x, coef) {
        median <- median(x)
        Q1 <- quantile(x, probs = .25)
        Q3 <- quantile(x, probs = .75)
        iqr <- Q3 - Q1

        lower_limit <- median - (iqr * coef)

        return(x < lower_limit)
    }

    hampel_upper <- function(x) {
        median <- median(x)
        difference <- abs(median - x)
        difference_median <- median(difference)

        upper_limit <- median + 3 * difference_median

        return(x > upper_limit)
    }

    hampel_lower <- function(x) {
        median <- median(x)
        difference <- abs(median - x)
        difference_median <- median(difference)

        lower_limit <- median - 3 * difference_median

        return(x < lower_limit)
    }

    for (col in names(df)) {
        if (is.numeric(df[[col]])) {
            if (ejection_type == 1) {
                df[iqr_lower(df[[col]], iqr_coef), col] <- min(df[!(iqr_lower(df[[col]], iqr_coef) | iqr_upper(df[[col]], iqr_coef)), col]) # nolint
                df[iqr_upper(df[[col]], iqr_coef), col] <- max(df[!(iqr_lower(df[[col]], iqr_coef) | iqr_upper(df[[col]], iqr_coef)), col]) # nolint
            } else if (ejection_type == 2) {
                df[hampel_lower(df[[col]]), col] <- min(df[!(hampel_lower(df[[col]]) | hampel_upper(df[[col]])), col]) # nolint
                df[hampel_upper(df[[col]]), col] <- max(df[!(hampel_upper(df[[col]]) | hampel_upper(df[[col]])), col]) # nolint
            }
        }
    }

    return(df)
}

fix_skipped <- function(data, nan_crit_val) {
    len <- length(names(data))
    drop <- list()
    for (i in 1:len) {
        if (is.numeric(data[[i]])) {
            nan_len <- length(data[, i][is.na(data[, i])])
            len_of_col <- length(data[, i])
            if (nan_len > len_of_col * nan_crit_val) {
                drop <- c(drop, names(data)[i])
            } else {
                data[, i][is.na(data[, i])] <- mean(data[, i], na.rm = T)
            }
        } else {
            nan_len <- length(data[, i][data[, i] == ""])
            len_of_col <- length(data[, i])
            if (nan_len > len_of_col * nan_crit_val) {
                drop <- c(drop, names(data)[i])
            } else {
                f <- factor(data[, i])
                data[, i][(data[, i]) == ""] <- levels(f)[2]
            }
        }
    }

    data <- data[, !(names(data) %in% drop)]
    return(data)
}

clean_data <- function(
    data,
    path_to_save,
    ejection_type = 1,
    nan_crit_val = 0.25,
    iqr_coef = 1) {
    print("START")
    print(data)

    data_new <- data[!duplicated(data), ]

    data_new <- fix_skipped(data_new, nan_crit_val)

    print("FIX SKIPPED:")
    print(data_new)

    data_new <- remove_ejections(data_new, iqr_coef, ejection_type)

    print("REMOVE EJECTIONS:")
    print(data_new)
    print("END")

    write.csv(data_new, path_to_save, row.names = FALSE)

    return(data_new)
}


draw_hists <- function(data1, data2, col) {
    if (is.numeric(data1[, col])) {
        # ggplot(data = data2) +
        #     geom_histogram(mapping = aes(x = data2[, col]), bins = 4)

        blue <- rgb(0, 0, 1, alpha = 0.5)
        red <- rgb(1, 0, 0, alpha = 0.5)

        width <- (max(data1[, col]) - min(data1[, col])) / 10
        count2 <- (max(data2[, col]) - min(data2[, col])) / width

        hist(data1[, col], prob = TRUE, col = blue, breaks = 9)
        hist(data2[, col], prob = TRUE, col = red, breaks = count2 - 1, add = TRUE)
        legend("topright",
            legend = c("old", "new"),
            col = c(blue, red),
            pch = 15
        )
    }
}

#########################################

df <- data.frame(
    fs = c(100, 90, NaN, 95, 1, 56, 56),
    ss = c(30, NaN, 45, 56, -11111, 56, 56),
    ts = c(52, 1240, 80, 98, 1, 56, 56),
    ffs = c(NaN, NaN, NaN, 65, 1, 56, 56),
    STRS = c("52", "40", "40", "", "aaa", 56, 56),
    stringsAsFactors = FALSE
)


directory <- getwd()
path <- paste(directory, "\\datasets\\avocado1.csv", sep = "")

df <- read.csv(path)

path_to_save <- paste(directory, "\\temp\\dc.csv", sep = "")

df_new <- clean_data(df, path_to_save, 1, 0.25, 1)

draw_hists(df, df_new, "AveragePrice")
